lua/ttt2/libraries/spawn.lua

```diff
@@ -0,0 +1,303 @@
+	local bottom, top = ply:GetHull()
+
+	return top - bottom
+end
+
+spawn = spawn or {}
+spawn.cachedSpawnEntities = spawn.cachedSpawnEntities or {}
+
+---
+-- Checks if a given spawn point is safe. Safe means that a player can spawn without
+-- being stuck. Entities which are passable are ignored by the check.
+-- @param Player ply The player entity that should spawn, this parameter is needed
+-- to make sure there is enough space for this specific playermodel.
+-- @param Vector pos The respawn position
+-- @param [default=false]boolean force Should the respawn be forced? This means killing other players that block this position
+-- @param [opt]table|Entity filter A table of entities or an entity that should be ignored by this check
+-- @return boolean Returns if the spawn point is safe
+-- @realm server
+function spawn.IsSpawnPointSafe(ply, pos, force, filter)
+	local sizePlayer = GetPlayerSize(ply)
+
+	if not util.IsInWorld(pos) then
+		return false
+	end
+
+	filter = istable(filter) and filter or {filter}
+
+	local posCenter = pos + Vector(0, 0, 0.5 * sizePlayer.z)
+
+	-- Make sure there is enough space around the player
+	local traceWalls = util.TraceHull({
+		start = posCenter,
+		endpos = posCenter,
+		mins = -0.5 * sizePlayer,
+		maxs = 0.5 * sizePlayer,
+		filter = function(ent)
+			if not IsValid(ent) or ent:IsPlayer() then
+				return false
+			end
+
+			if ent:HasPassableCollisionGrup() then
+				return false
+			end
+
+			return true
+		end,
+		mask = MASK_SOLID
+	})
+
+	if traceWalls.Hit then
+		return false
+	end
+
+	-- make sure the new position is on the ground
+	local traceGround = util.TraceLine({
+		start = posCenter,
+		endpos = posCenter - Vector(0, 0, sizePlayer.z),
+		filter = player.GetAll(),
+		mask = MASK_SOLID
+	})
+
+	if not traceGround.Hit then
+		return false
+	end
+
+	local blockingEntities = ents.FindInBox(
+		pos + Vector(-0.5 * sizePlayer.x, -0.5 * sizePlayer.y, 0),
+		pos + sizePlayer
+	)
+
+	for i = 1, #blockingEntities do
+		local blockingEntity = blockingEntities[i]
+
+		if not IsValid(blockingEntity) or not blockingEntity:IsPlayer()
+			or not blockingEntity:IsTerror() or not blockingEntity:Alive()
+		then continue end
+
+		if table.HasValue(filter, blockingEntity) then continue end
+
+		if force then
+			blockingEntity:Kill()
+		else
+			return false
+		end
+	end
+
+	return true
+end
+
+---
+-- Returns a selection of points around the given position.
+-- @param Player ply The player entity that should spawn, this parameter is needed
+-- to make sure there is enough space for this specific playermodel.
+-- @param Vector pos The given position
+-- @return table A table of position vectors
+-- @realm server
+function spawn.GetSpawnPointsAroundSpawn(ply, pos)
+	local sizePlayer = GetPlayerSize(ply)
+
+	if not pos then return {} end
+
+	local positions = {}
+
+	for i = 1, #spawnPointVariations do
+		positions[#positions + 1] = pos + spawnPointVariations[i] * Vector(sizePlayer.x, sizePlayer.y, 0) * 1.5
```

Comment:
Better using `positions[i] = ...`

Reply:
The vector `Vector(sizePlayer.x, sizePlayer.y, 0)` should be stored outside this loop. You can multiply the 1.5 as well